src/core/utils/debug_uart.h

```diff
@@ -0,0 +1,186 @@
+#define PLATFORM_DEBUG_UART_H
+/*
+ *  Copyright (c) 2017, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements a very basic debug uart useful for logging
+ *   messages when the primary uart is already engaged (example the NCP)
+ *   and encapsulating logs is not practical.
+ *
+ * Implimentation Notes:
+ *
+ * Only 3 functions are required to be implimented by the platform.
+ * see/search for the word MUST in the comments below.
+ *
+ * The other functions (without MUST in the comment) are WEAK symbols
+ * that can optionally be overridden by the platform or the application.
+ *
+ * many of the other functions are simply convience functions to
+ * aid a developer during their normal course of work and are not
+ * intended to be present, or used in production system.
+ */
+
+#if __cplusplus
+extern "C" {
+#endif
+
+    /**
+     * Standard printf() to the debug uart with no log decoration.
+     *
+     * @param fmt[in] printf formatter text
+     *
+     * This is a debug convience function that is not intended to be
+     * used in anything other then "debug scenarios" by a developer.
+     *
+     * lf -> cr/lf mapping is automatically handled via otPlatDebugUart_putchar()
+     *
+     * @sa otPlatDebugUart_vprintf() for limitations
+     *
+     * This is a WEAK symbol that can easily be overridden as needed.
+     */
+    void otPlatDebugUart_printf(const char *fmt, ...);
```

Comment:
**PART 1**

First: These are never suggested as a replacement for, or use in normal 'release code'.
Developers should use the standard log functions when and where they are usable.

abtink>> If so, why not use the existing otLog<Level><Module>() APIs and just map otPlatLog to use a sperate debug/log UART (which may be separate channel/uart from what NCP uses)...

That is exactly what the code is doing here, just look at the call flow (below)

That said, the developer is free to use those platform level functions if they choose to.
perhaps  have made it more confusing by pointing out that somebody *can* use these.

**START HERE:**  The `otLog<Level><Module>()`  macro - in the DebugUart case ...

The macro expands to:  `otPlatDebugUart_otPlatLog()` 

https://github.com/DuaneEllis-TI/openthread/blob/debug-uart/src/core/utils/debug_uart.cpp#L107

That function ... 

Invokes: `otPlatDebugUart_printf()` and 
Invokes: `otPlatDebugUart_vprintf()` 
Invokes: `otPlatDebugUart_putchar()` for the terminal newline.

They in turn ... invokes others ... and they need prototypes

Example:   `otPlatDebugUart_puts_no_nl()` 

https://github.com/DuaneEllis-TI/openthread/blob/debug-uart/src/core/utils/debug_uart.cpp#L78

Example:  `otPlatDebugUart_putchar()` 

https://github.com/DuaneEllis-TI/openthread/blob/debug-uart/src/core/utils/debug_uart.cpp#L195

Example: `otPlatDebugUart_putchar_raw()`  a platform specific function

https://github.com/DuaneEllis-TI/openthread/blob/debug-uart/examples/platforms/cc2652/uart.c#L329

**PART 2**

But you asked about PROS & CONS of this?

Rather then have every platform implement the common stuff... 

Example, Nortic has a solution tied 100% to the SEGGER debugger, see this as an example:

https://github.com/openthread/openthread/blob/master/examples/platforms/nrf52840/logging.c

It would be better to refactor something like that and make it available to all platforms, that is what I am doing here.

**PART 3**

That said, we could hide those internal functions 

But by exposing these functions there are a few underlying features that are helpful here.

1) To keep the platform layer super simple, the best API at the platform specific layer 3 functions:  uart write byte,  uart kbhit, and uart get byte.  that's the minimal requirement for an platform.

This helps with the above quit a bit.

2) You'll notice that all of these functions are "weak" - for a reason.

For some platforms, and some debug situations - Forcing all debug to go through "putchar" is at times a horribly slow thing :-( but it is simple and works in the common case.

A better solution on Cortex platforms is the SWO trace feature - different vendors have different methods of exposing the feature, that's why there are a few layers of code exposed the silicon vendor can choose which one they want to connect at - by overriding the specific WEAK symbol.

This is sort of what Nortic is doing with the calls to the SEGGER debugger functions.

3) In the *NON*  log scheme - when you are debugging crashes - these functions are helpful in another way.

The standard LOG functions require and use printf() - In the embedded world there are times when for example, when printf() causes problems.  For example some versions of printf() require a heap (malloc/free) that is functional, or everything else has been scribbled on because of the bug  you are hunting down.

Technically to meet the goal of the debug uart the other functions are technically not required, but - they are exposed for the reasons above - there is nothing stopping somebody from using them - which leads me to the next reason.

4) Often when debugging something things "lock up" - and end at a while(1) loop. that's nice, but hard to debug. Even harder if the CPU ended up at the FaultISR vector.  even harder if the debugger was not attached and ready to go, sometimes plugging the debugger in will reset the board. GRRR!!!!

Think about a test farm with 30 devices running the compliance test suite

Why not convert the wile(1) loop into something that continuously dumps the cpu registers and the stack, all you need to do is connect the serial cable.

Take a look at this, this is the basis for CortexM parts

       http://www.freertos.org/Debugging-Hard-Faults-On-Cortex-M-Microcontrollers.html

The challenge here is this: You cannot use printf() because the infrastructure it requires is toast.  

Writing a simple print hex values routine is simple given these lower level functions, and won't take up much code

Here's an example from ARM7 tdmi in some stuff I wrote years ago.

Starting at the data abort handler (equal to the FAULT handler in CortexM series)

https://sourceforge.net/p/lostarm/code/HEAD/tree/trunk/chips/generic_arm/source/default_da_handler.c

Which calls a common crash handler:

https://sourceforge.net/p/lostarm/code/HEAD/tree/trunk/chips/generic_arm/source/common_xxx_handler.c

Which then dumps cpu registers

https://sourceforge.net/p/lostarm/code/HEAD/tree/trunk/chips/generic_arm/source/reg-dump.c

And a small stack dump (the ARM7tdmi provides a stack frame pointer, which makes printing a stack backtrace simple)

https://sourceforge.net/p/lostarm/code/HEAD/tree/trunk/chips/generic_arm/source/stackdump.c

-Duane.